File: backend/app/services/tags.py
from __future__ import annotations

from datetime import datetime
import re

from fastapi import HTTPException, status
from sqlalchemy import case, func, select
from sqlalchemy.orm import Session, selectinload

from app.models.book import Book, BookAuthor, BookTitle
from app.models.tag import BookTagVote, Tag, TagRequest, TagRequestStatus
from app.models.user import User, UserBook
from app.schemas.book import BookSummary
from app.schemas.tag import (
    BookTagAggregate,
    TagCreate,
    TagDetail,
    TaggedBookSummary,
    TagRead,
    TagRequestCreate,
    TagUpdate,
)


def normalize_tag_slug(name: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", name.strip().lower()).strip("-")
    if not slug:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Tag name must include letters or numbers")
    return slug


def _clean_name(name: str) -> str:
    cleaned = " ".join(name.split())
    if not cleaned:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Tag name cannot be empty")
    return cleaned


def _get_parent_or_404(db: Session, parent_id: int | None) -> Tag | None:
    if parent_id is None:
        return None
    parent = db.get(Tag, parent_id)
    if parent is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Parent tag not found")
    return parent


def _require_slug_available(db: Session, slug: str, *, ignore_tag_id: int | None = None) -> None:
    statement = select(Tag).where(Tag.slug == slug)
    if ignore_tag_id is not None:
        statement = statement.where(Tag.id != ignore_tag_id)
    if db.scalar(statement) is not None:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="A tag with this name already exists")


def _require_pending_request_slug_available(db: Session, slug: str) -> None:
    tag_request = db.scalar(
        select(TagRequest).where(
            TagRequest.proposed_slug == slug,
            TagRequest.status == TagRequestStatus.PENDING,
        )
    )
    if tag_request is not None:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="A pending tag request with this name already exists")


def _require_no_parent_cycle(db: Session, tag: Tag, parent_id: int | None) -> None:
    ancestor_id = parent_id
    seen_ids: set[int] = set()
    while ancestor_id is not None:
        if ancestor_id == tag.id:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Tag parent cannot create a cycle")
        if ancestor_id in seen_ids:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Tag tree already contains a cycle")
        seen_ids.add(ancestor_id)
        ancestor = db.get(Tag, ancestor_id)
        if ancestor is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Parent tag not found")
        ancestor_id = ancestor.parent_id


def create_tag(db: Session, payload: TagCreate, *, current_user: User) -> Tag:
    _get_parent_or_404(db, payload.parent_id)
    name = _clean_name(payload.name)
    slug = normalize_tag_slug(name)
    _require_slug_available(db, slug)

    tag = Tag(
        name=name,
        slug=slug,
        description=payload.description.strip(),
        aliases=payload.aliases.strip(),
        default_spoiler_level=payload.default_spoiler_level,
        is_applicable=payload.is_applicable,
        parent_id=payload.parent_id,
        created_by_id=current_user.id,
        approved_by_id=current_user.id,
    )
    db.add(tag)
    db.commit()
    db.refresh(tag)
    return tag


def update_tag(db: Session, tag_id: int, payload: TagUpdate) -> Tag:
    tag = db.get(Tag, tag_id)
    if tag is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")

    updates = payload.model_dump(exclude_unset=True)
    if "parent_id" in updates:
        _require_no_parent_cycle(db, tag, payload.parent_id)
        _get_parent_or_404(db, payload.parent_id)
        tag.parent_id = payload.parent_id
    if payload.name is not None:
        name = _clean_name(payload.name)
        slug = normalize_tag_slug(name)
        _require_slug_available(db, slug, ignore_tag_id=tag.id)
        tag.name = name
        tag.slug = slug
    if payload.description is not None:
        tag.description = payload.description.strip()
    if payload.aliases is not None:
        tag.aliases = payload.aliases.strip()
    if payload.default_spoiler_level is not None:
        tag.default_spoiler_level = payload.default_spoiler_level
    if payload.is_applicable is not None:
        tag.is_applicable = payload.is_applicable

    db.commit()
    db.refresh(tag)
    return tag


def create_tag_request(db: Session, payload: TagRequestCreate, *, current_user: User) -> TagRequest:
    _get_parent_or_404(db, payload.parent_id)
    name = _clean_name(payload.proposed_name)
    slug = normalize_tag_slug(name)
    _require_slug_available(db, slug)
    _require_pending_request_slug_available(db, slug)

    tag_request = TagRequest(
        requested_by_id=current_user.id,
        parent_id=payload.parent_id,
        proposed_name=name,
        proposed_slug=slug,
        description=payload.description.strip(),
        submitter_note=payload.submitter_note.strip(),
    )
    db.add(tag_request)
    db.commit()
    db.refresh(tag_request)
    return tag_request


def approve_tag_request(
    db: Session,
    tag_request: TagRequest,
    *,
    reviewer: User,
    reviewer_note: str = "",
) -> None:
    if tag_request.status != TagRequestStatus.PENDING:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Tag request has already been reviewed")

    _get_parent_or_404(db, tag_request.parent_id)
    _require_slug_available(db, tag_request.proposed_slug)
    tag = Tag(
        name=tag_request.proposed_name,
        slug=tag_request.proposed_slug,
        description=tag_request.description,
        parent_id=tag_request.parent_id,
        created_by_id=tag_request.requested_by_id,
        approved_by_id=reviewer.id,
    )
    db.add(tag)
    db.flush()

    tag_request.status = TagRequestStatus.APPROVED
    tag_request.reviewed_by_id = reviewer.id
    tag_request.reviewer_note = reviewer_note
    tag_request.reviewed_at = datetime.now().astimezone()
    tag_request.created_tag_id = tag.id


def reject_tag_request(
    db: Session,
    tag_request: TagRequest,
    *,
    reviewer: User,
    reviewer_note: str = "",
) -> None:
    if tag_request.status != TagRequestStatus.PENDING:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Tag request has already been reviewed")

    tag_request.status = TagRequestStatus.REJECTED
    tag_request.reviewed_by_id = reviewer.id
    tag_request.reviewer_note = reviewer_note
    tag_request.reviewed_at = datetime.now().astimezone()


def tag_ancestors(tags_by_id: dict[int, Tag], tag: Tag) -> list[Tag]:
    ancestors: list[Tag] = []
    parent_id = tag.parent_id
    seen_ids: set[int] = set()
    while parent_id is not None and parent_id not in seen_ids:
        seen_ids.add(parent_id)
        parent = tags_by_id.get(parent_id)
        if parent is None:
            break
        ancestors.append(parent)
        parent_id = parent.parent_id
    ancestors.reverse()
    return ancestors


def _book_summary_options():
    return (
        selectinload(Book.authors).joinedload(BookAuthor.author),
        selectinload(Book.titles).joinedload(BookTitle.language),
        selectinload(Book.covers),
    )


def _book_rating_subquery():
    return (
        select(
            UserBook.book_id.label("book_id"),
            func.avg(UserBook.rating).label("average_rating"),
            func.count(UserBook.rating).label("rating_count"),
        )
        .group_by(UserBook.book_id)
        .subquery()
    )


def _book_summary_with_rating(book: Book, average_rating: float | None, rating_count: int | None) -> BookSummary:
    return BookSummary.model_validate(book).model_copy(
        update={
            "average_rating": round(float(average_rating), 2) if average_rating is not None else None,
            "rating_count": int(rating_count or 0),
        }
    )


def get_tag_detail(
    db: Session,
    tag_id: int,
    *,
    show_all: bool = False,
    show_spoilers: bool = False,
) -> TagDetail:
    tag = db.get(Tag, tag_id)
    if tag is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")

    positive_vote = case((BookTagVote.vote > 0, 1), else_=0)
    downvote = case((BookTagVote.vote < 0, 1), else_=0)
    positive_rating = case((BookTagVote.vote > 0, BookTagVote.vote), else_=None)
    spoiler_vote = case(((BookTagVote.vote > 0) & (BookTagVote.spoiler_level > 0), 1), else_=0)
    positive_spoiler_level = case((BookTagVote.vote > 0, BookTagVote.spoiler_level), else_=0)
    tag_book_aggregate = (
        select(
            BookTagVote.book_id.label("book_id"),
            func.sum(BookTagVote.vote).label("score"),
            func.count(BookTagVote.id).label("vote_count"),
            func.sum(positive_vote).label("positive_vote_count"),
            func.sum(downvote).label("downvote_count"),
            func.sum(spoiler_vote).label("spoiler_vote_count"),
            func.max(positive_spoiler_level).label("aggregate_spoiler_level"),
            func.avg(positive_rating).label("average_positive_rating"),
        )
        .where(BookTagVote.tag_id == tag_id)
        .group_by(BookTagVote.book_id)
        .subquery()
    )
    rating_subquery = _book_rating_subquery()
    statement = (
        select(
            Book,
            tag_book_aggregate.c.score,
            tag_book_aggregate.c.vote_count,
            tag_book_aggregate.c.positive_vote_count,
            tag_book_aggregate.c.downvote_count,
            tag_book_aggregate.c.spoiler_vote_count,
            tag_book_aggregate.c.aggregate_spoiler_level,
            tag_book_aggregate.c.average_positive_rating,
            rating_subquery.c.average_rating,
            rating_subquery.c.rating_count,
        )
        .join(tag_book_aggregate, tag_book_aggregate.c.book_id == Book.id)
        .outerjoin(rating_subquery, rating_subquery.c.book_id == Book.id)
        .options(*_book_summary_options())
        .order_by(
            tag_book_aggregate.c.score.desc(),
            tag_book_aggregate.c.average_positive_rating.desc().nullslast(),
            tag_book_aggregate.c.vote_count.desc(),
            Book.id.asc(),
        )
    )

    books: list[TaggedBookSummary] = []
    for (
        book,
        score,
        vote_count,
        positive_vote_count,
        downvote_count,
        spoiler_vote_count,
        aggregate_spoiler_level,
        average_positive_rating,
        average_rating,
        rating_count,
    ) in db.execute(statement).all():
        score_value = int(score or 0)
        aggregate_spoiler_level_value = int(aggregate_spoiler_level or 0)
        if score_value <= 0 and not show_all:
            continue
        if aggregate_spoiler_level_value > 0 and not show_spoilers:
            continue

        books.append(
            TaggedBookSummary(
                book=_book_summary_with_rating(book, average_rating, rating_count),
                relevance_score=score_value,
                vote_count=int(vote_count or 0),
                positive_vote_count=int(positive_vote_count or 0),
                downvote_count=int(downvote_count or 0),
                spoiler_vote_count=int(spoiler_vote_count or 0),
                aggregate_spoiler_level=aggregate_spoiler_level_value,
                average_positive_rating=(
                    round(float(average_positive_rating), 2)
                    if average_positive_rating is not None
                    else None
                ),
            )
        )

    children = db.scalars(select(Tag).where(Tag.parent_id == tag_id).order_by(Tag.name.asc(), Tag.id.asc())).all()
    return TagDetail(
        **TagRead.model_validate(tag).model_dump(),
        parent=TagRead.model_validate(tag.parent) if tag.parent is not None else None,
        children=[TagRead.model_validate(child) for child in children],
        books=books,
    )


def list_book_tag_aggregates(
    db: Session,
    book_id: int,
    *,
    current_user: User | None = None,
    show_all: bool = False,
    show_spoilers: bool = False,
) -> list[BookTagAggregate]:
    tags_by_id = {tag.id: tag for tag in db.scalars(select(Tag)).all()}
    current_user_votes: dict[int, BookTagVote] = {}

    if current_user is not None:
        current_user_votes = {
            vote.tag_id: vote
            for vote in db.scalars(
                select(BookTagVote).where(
                    BookTagVote.book_id == book_id,
                    BookTagVote.user_id == current_user.id,
                )
            ).all()
        }

    positive_vote = case((BookTagVote.vote > 0, 1), else_=0)
    downvote = case((BookTagVote.vote < 0, 1), else_=0)
    positive_rating = case((BookTagVote.vote > 0, BookTagVote.vote), else_=None)
    spoiler_vote = case(((BookTagVote.vote > 0) & (BookTagVote.spoiler_level > 0), 1), else_=0)
    positive_spoiler_level = case((BookTagVote.vote > 0, BookTagVote.spoiler_level), else_=0)
    statement = (
        select(
            Tag,
            func.sum(BookTagVote.vote).label("score"),
            func.count(BookTagVote.id).label("vote_count"),
            func.sum(positive_vote).label("positive_vote_count"),
            func.sum(downvote).label("downvote_count"),
            func.sum(spoiler_vote).label("spoiler_vote_count"),
            func.max(positive_spoiler_level).label("aggregate_spoiler_level"),
            func.avg(positive_rating).label("average_positive_rating"),
        )
        .join(BookTagVote, BookTagVote.tag_id == Tag.id)
        .where(BookTagVote.book_id == book_id)
        .group_by(Tag.id)
        .order_by(
            func.sum(BookTagVote.vote).desc(),
            func.count(BookTagVote.id).desc(),
            Tag.name.asc(),
        )
    )

    aggregates: list[BookTagAggregate] = []
    for (
        tag,
        score,
        vote_count,
        positive_vote_count,
        downvote_count,
        spoiler_vote_count,
        aggregate_spoiler_level,
        average_positive_rating,
    ) in db.execute(statement).all():
        score_value = int(score or 0)
        aggregate_spoiler_level_value = int(aggregate_spoiler_level or 0)
        if score_value <= 0 and not show_all:
            continue
        if aggregate_spoiler_level_value > 0 and not show_spoilers:
            continue

        current_user_vote = current_user_votes.get(tag.id)
        current_user_spoiler_level = current_user_vote.spoiler_level if current_user_vote is not None else None

        aggregates.append(
            BookTagAggregate(
                tag=TagRead.model_validate(tag),
                ancestors=[TagRead.model_validate(ancestor) for ancestor in tag_ancestors(tags_by_id, tag)],
                score=score_value,
                vote_count=int(vote_count or 0),
                positive_vote_count=int(positive_vote_count or 0),
                downvote_count=int(downvote_count or 0),
                spoiler_vote_count=int(spoiler_vote_count or 0),
                aggregate_spoiler_level=aggregate_spoiler_level_value,
                is_spoiler=aggregate_spoiler_level_value > 0,
                average_positive_rating=(
                    round(float(average_positive_rating), 2)
                    if average_positive_rating is not None
                    else None
                ),
                current_user_vote=current_user_vote.vote if current_user_vote is not None else None,
                current_user_spoiler_level=current_user_spoiler_level,
                current_user_spoiler=(
                    current_user_spoiler_level > 0
                    if current_user_spoiler_level is not None
                    else None
                ),
            )
        )
    return aggregates

File: frontend/src/api/tags.ts
import { apiRequest } from './client'
import type { BookSummary } from './books'

export type TagVoteValue = -1 | 1 | 2 | 3
export type SpoilerLevel = 0 | 1 | 2
export type TagRequestStatus = 'pending' | 'approved' | 'rejected'

export interface TagRead {
  id: number
  name: string
  slug: string
  description: string
  aliases: string
  default_spoiler_level: SpoilerLevel
  is_applicable: boolean
  parent_id?: number | null
  created_by_id?: number | null
  approved_by_id?: number | null
  created_at: string
  updated_at: string
}

export interface TagTreeNode extends TagRead {
  children: TagTreeNode[]
}

export interface BookTagAggregate {
  tag: TagRead
  ancestors: TagRead[]
  score: number
  vote_count: number
  positive_vote_count: number
  downvote_count: number
  spoiler_vote_count: number
  aggregate_spoiler_level: SpoilerLevel
  average_positive_rating?: number | null
  current_user_vote?: number | null
  current_user_spoiler_level?: SpoilerLevel | null
}

export interface TaggedBookSummary {
  book: BookSummary
  relevance_score: number
  vote_count: number
  positive_vote_count: number
  downvote_count: number
  spoiler_vote_count: number
  aggregate_spoiler_level: SpoilerLevel
  average_positive_rating?: number | null
}

export interface TagDetail extends TagRead {
  parent?: TagRead | null
  children: TagRead[]
  books: TaggedBookSummary[]
}

export interface BookTagVoteRead {
  id: number
  book_id: number
  tag_id: number
  user_id: number
  vote: TagVoteValue
  spoiler_level: SpoilerLevel
  created_at: string
  updated_at: string
}

export interface CreateTagRequestInput {
  proposed_name: string
  parent_id?: number | null
  description?: string
  submitter_note?: string
}

export interface CreateTagInput {
  name: string
  parent_id?: number | null
  description?: string
  aliases?: string
  default_spoiler_level?: SpoilerLevel
  is_applicable?: boolean
}

export interface TagRequestRead {
  id: number
  requested_by_id: number
  reviewed_by_id?: number | null
  parent_id?: number | null
  proposed_name: string
  proposed_slug: string
  description: string
  submitter_note: string
  reviewer_note: string
  status: TagRequestStatus
  created_tag_id?: number | null
  created_at: string
  updated_at: string
  reviewed_at?: string | null
}

export function flattenTagTree(tags: TagTreeNode[]): Array<TagRead & { path: string }> {
  const flattened: Array<TagRead & { path: string }> = []

  function visit(tag: TagTreeNode, ancestors: string[]) {
    const pathParts = [...ancestors, tag.name]
    flattened.push({ ...tag, path: pathParts.join(' > ') })
    tag.children.forEach(child => visit(child, pathParts))
  }

  tags.forEach(tag => visit(tag, []))
  return flattened
}

export async function listTags(): Promise<TagTreeNode[]> {
  return apiRequest<TagTreeNode[]>('/api/v1/tags', {
    fallbackErrorMessage: 'Unable to load tags.',
  })
}

export async function getTag(
  tagId: string | number,
  showAll = false,
  showSpoilers = false,
): Promise<TagDetail> {
  const params = new URLSearchParams()
  if (showAll) {
    params.set('show_all', 'true')
  }
  if (showSpoilers) {
    params.set('show_spoilers', 'true')
  }

  return apiRequest<TagDetail>(`/api/v1/tags/${tagId}${params.size ? `?${params.toString()}` : ''}`, {
    fallbackErrorMessage: 'Unable to load this tag.',
  })
}

export async function listBookTags(
  bookId: string | number,
  accessToken?: string | null,
  showAll = false,
  showSpoilers = false,
): Promise<BookTagAggregate[]> {
  const params = new URLSearchParams()
  if (showAll) {
    params.set('show_all', 'true')
  }
  if (showSpoilers) {
    params.set('show_spoilers', 'true')
  }

  return apiRequest<BookTagAggregate[]>(`/api/v1/books/${bookId}/tags${params.size ? `?${params.toString()}` : ''}`, {
    accessToken: accessToken ?? undefined,
    fallbackErrorMessage: 'Unable to load book tags.',
  })
}

export async function voteBookTag(
  bookId: string | number,
  tagId: string | number,
  vote: TagVoteValue,
  spoilerLevel: SpoilerLevel,
  accessToken: string,
): Promise<BookTagVoteRead> {
  return apiRequest<BookTagVoteRead>(`/api/v1/books/${bookId}/tags/${tagId}/vote`, {
    accessToken,
    method: 'PUT',
    body: { vote, spoiler_level: spoilerLevel },
    fallbackErrorMessage: 'Unable to save this tag vote.',
  })
}

export async function clearBookTagVote(
  bookId: string | number,
  tagId: string | number,
  accessToken: string,
): Promise<void> {
  await apiRequest<void>(`/api/v1/books/${bookId}/tags/${tagId}/vote`, {
    accessToken,
    method: 'DELETE',
    fallbackErrorMessage: 'Unable to clear this tag vote.',
  })
}

export async function createTagRequest(
  accessToken: string,
  payload: CreateTagRequestInput,
): Promise<TagRequestRead> {
  return apiRequest<TagRequestRead>('/api/v1/tag-requests', {
    accessToken,
    method: 'POST',
    body: payload,
    fallbackErrorMessage: 'Unable to submit this tag request.',
  })
}

export async function createAdminTag(
  accessToken: string,
  payload: CreateTagInput,
): Promise<TagRead> {
  return apiRequest<TagRead>('/api/v1/admin/tags', {
    accessToken,
    method: 'POST',
    body: payload,
    fallbackErrorMessage: 'Unable to create this tag.',
  })
}

export async function listMyTagRequests(accessToken: string): Promise<TagRequestRead[]> {
  return apiRequest<TagRequestRead[]>('/api/v1/tag-requests/me', {
    accessToken,
    fallbackErrorMessage: 'Unable to load your tag requests.',
  })
}

export async function listAdminTagRequests(
  accessToken: string,
  status?: TagRequestStatus,
): Promise<TagRequestRead[]> {
  const params = new URLSearchParams()
  if (status) {
    params.set('request_status', status)
  }

  return apiRequest<TagRequestRead[]>(`/api/v1/admin/tag-requests${params.size ? `?${params.toString()}` : ''}`, {
    accessToken,
    fallbackErrorMessage: 'Unable to load tag requests.',
  })
}

export async function reviewTagRequest(
  accessToken: string,
  tagRequestId: string | number,
  payload: { status: Exclude<TagRequestStatus, 'pending'>; reviewer_note?: string },
): Promise<TagRequestRead> {
  return apiRequest<TagRequestRead>(`/api/v1/admin/tag-requests/${tagRequestId}/review`, {
    accessToken,
    method: 'POST',
    body: payload,
    fallbackErrorMessage: 'Unable to review this tag request.',
  })
}
